test/e2e/reconciler/namespace/controller_test.go

```diff
@@ -275,56 +187,29 @@ func scheduledMatcher(target string) matcher {
 	}
 }
 
-func scheduledAnywhereMatcher() matcher {
-	return func(object *corev1.Namespace) error {
-		if isNamespaceUnschedulable(object) {
-			return fmt.Errorf("expected a scheduled workspace, got status.conditions: %#v", object.Status.Conditions)
-		}
-		return nil
-	}
-}
-
-type matcher func(object *corev1.Namespace) error
-
-func expectNextEvent(w watch.Interface, expectType watch.EventType, matcher matcher, duration time.Duration) (*corev1.Namespace, error) {
-	event, err := nextEvent(w, duration)
-	if err != nil {
-		return nil, err
-	}
-	if expectType != event.Type {
-		return nil, fmt.Errorf("got incorrect watch event type: %v != %v\n", expectType, event.Type)
-	}
-	workspace, ok := event.Object.(*corev1.Namespace)
-	if !ok {
-		return nil, fmt.Errorf("got %T, not a Workspace", event.Object)
-	}
-	if err := matcher(workspace); err != nil {
-		return workspace, err
-	}
-	return workspace, nil
-}
+type registerNamespaceExpectation func(seed *corev1.Namespace, expectation namespaceExpectation) error
 
-func nextEvent(w watch.Interface, duration time.Duration) (watch.Event, error) {
-	stopTimer := time.NewTimer(duration)
-	defer stopTimer.Stop()
-	select {
-	case event, ok := <-w.ResultChan():
-		if !ok {
-			return watch.Event{}, errors.New("watch closed unexpectedly")
-		}
-		return event, nil
-	case <-stopTimer.C:
-		return watch.Event{}, errors.New("timed out waiting for event")
+func expectNamespaces(ctx context.Context, t framework.TestingTInterface, client kubernetes.Interface) (registerNamespaceExpectation, error) {
+	informerFactory := informers.NewSharedInformerFactory(client, 0)
+	informer := informerFactory.Core().V1().Namespaces()
+	expecter := framework.NewExpecter(informer.Informer())
+	informerFactory.Start(ctx.Done())
+	if !cache.WaitForNamedCacheSync(t.Name(), ctx.Done(), informer.Informer().HasSynced) {
+		return nil, errors.New("failed to wait for caches to sync")
 	}
-}
-
-const conditionTypeUnschedulable = "Unschedulable"
-
-func isNamespaceUnschedulable(ns *corev1.Namespace) bool {
-	for _, cond := range ns.Status.Conditions {
-		if cond.Type == conditionTypeUnschedulable {
-			return cond.Status == corev1.ConditionTrue
+	return func(seed *corev1.Namespace, expectation namespaceExpectation) error {
+		key, err := cache.MetaNamespaceKeyFunc(seed)
+		if err != nil {
+			return err
 		}
-	}
-	return false
+		return expecter.ExpectBefore(ctx, func(ctx context.Context) (done bool, err error) {
+			current, err := informer.Lister().Get(key)
+			if err != nil {
+				// Retry on all errors
```

Comment:
Depends on your intent but 403 -> 502 -> 200 would IMHO be a valid failure for the test, but this code would swallow the first broken responses as long as it converges?

Reply:
I'm not sure I follow. How would this test be able to differentiate hard failure vs retryable failure from those codes? And what would be the value of doing so?

Reply:
We're checking to response from the lister periodically (after events are received by the watcher) - if the intent of the test is to say "an object eventually exists with some state", ignoring 404 is strictly correct. Ignoring *all* errors may be more of a stretch, or it may not be - depends on what you intend.

Reply:
I'm afraid I'm still missing your point. How do I determine whether an arbitrary error returned by the lister is fatal or not? And is there harm in assuming retry given that any pathological failure will hit a timeout and be reported as a test failure?

Reply:
Fortunately for lister.Get, the only error you can get (today, at least) is a NotFound error.

I see what Steve is asking, but I don't think it's anything we can deal with (Maru's point about not knowing the severity of an error), and I think eventual convergence + hitting a timeout if we don't is good enough. WDYT?

Reply:
I think my point is that "how you determine" is the code you write in this line - the other examples explicitly choose to ignore 404 and let state converge, but it is with your err check here that you state your preferred intent for the test - whether it is or is not fatal. It might be expected that a 403 or a 502 would be unexpected and fatal, or it might not - I was mostly making sure that we were clear in the fact that your choice here says that a string of requests 403 -> 502 -> 200 was OK for the client to see. Good point on what the lister can ever return though, it's clearly moot here